### PR TITLE
Possible to join from the WebUI

### PR DESCRIPTION
**Pull request by nicemicro**
This PR contains the following:

- parsing JSON that is received from the webUI
- sending JSON to the webUI from Godot
- joining the game if the game status is "lobby"

### CODE REVIEW

File: controlpad/controller/controlpad.js [hunk 1]
```diff
@@ -44,4 +46,21 @@ export function send_controlpad_message(msg) {
     ws.send(msg);
 }
 
+export function send_datum(msg) {
```
Comment (safwaan-menk) [inline controlpad/controller/controlpad.js]: send_datum is the same as send_controlpad_message

File: controlpad/controller/controlpad.js [hunk 2]
```diff
@@ -30,9 +29,12 @@ ws.onopen = (_event) => {
             }
         }
         else {
-            let msg = event.data;
-            console.log("received message" + '<' + msg + '>');
-            console.log('^^ WASNT EXPECTING TO RECV ANY MSGS ^^');
+            const data = JSON.stringify(event.data);
```
Comment (safwaan-menk) [inline controlpad/controller/controlpad.js]: Ideally this file shouldn't me modified. 

This file will eventually become just a symbolic link on the console and any custom changes here won't be reflected. 

I can export data and create a handle message in the app.js if that's better.

File: controlpad/controller/controlpad.js [hunk 3]
```diff
@@ -44,4 +46,21 @@ export function send_controlpad_message(msg) {
...
+export const MsgType = {
+    "SUBMIT_NAME": 'SUBMIT_NAME'
+}
+
+/**
+ * Send message via websocket
+ * @param {string} type
+ * @param {Record<string, any>} payload
+ */
+export function send_msg(type, payload) {
```
Comment (safwaan-menk) [inline controlpad/controller/controlpad.js]: Same as first comment about this file becoming a symlink. Handle this in app.js or another file

File: controlpad/controller/required.js [hunk 4]
```diff
@@ -95,45 +95,13 @@ function togglePopup() {
     overlay.style.display = isHidden ? 'block' : 'none';
 }
 
-// ------------------------- Orientation Check ----------------------------------
```
Comment (safwaan-menk) [inline controlpad/controller/required.js:98]: These changes are fine. This file will also be a symlink pointing to a local required.js on the machine, but I was going to remove this code from here anyways so it's okay to be deleted

File: godot_code/main.tscn [hunk 5]
```diff
@@ -22,45 +16,7 @@ script = ExtResource("1_heqex")
...
-size_flags_horizontal = 3
-
-[node name="ExitButton" type="Button" parent="MainWindow/Columns/Right"]
-layout_mode = 2
-size_flags_horizontal = 8
-size_flags_vertical = 8
-text = "Exit"
 
 [connection signal="message_received" from="GameNiteControlpads" to="." method="_on_game_nite_controlpads_message_received"]
-[connection signal="pressed" from="MainWindow/Columns/Right/ExitButton" to="." method="_on_exit_button_pressed"]
```
Comment (safwaan-menk) [inline godot_code/main.tscn:66]: User has an exit button by default on their controller. No need to add one in the game unless you are using it just for debugging purposes

File: godot_code/start_window.gd [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+	var menuScene = preload("res://main_menu.tscn")
+	for child in menuPlace.get_children():
+		child.queue_free()
+	var menuNode: VBoxContainer = menuScene.instantiate()
+	lobbyNode = null
+	menuNode.connect("startButtonPressed", openGameJoin)
+	menuPlace.add_child(menuNode)
+	emit_signal("stateChanged", "mainMenu")
+
+func _on_exit_button_pressed():
```
Comment (safwaan-menk) [inline godot_code/start_window.gd:34]: When the game runs through the console software it runs as a child process. The console software kills the the child process when it receives an exit message so the game itself should not handle the exit

**Review by safwaan-menk — CHANGES_REQUESTED** [review]
- Code works from Godot.
- Address the small review comments and then you can approve and merge
- Controller is not fully in sync with the game yet. For example, I can't access the start menu button from the controller yet

Design decision:
- Should the menu and lobby be the same screen? That way when a client connects then they can just enter their name and join the lobby, rather than having an extra set of steps. The menu options can be present at the bottom horizontally like 

----- Start ----- Rules ---- Credits -----

Maybe all of those menu options can be handled by the phone controller also, just so there is no dpad required for the game

File: godot_code/main.tscn (hunk 5 above)
Comment (nicemicro) [reply godot_code/main.tscn:66]: it is for debugging at the moment.

File: godot_code/start_window.gd (hunk 6 above)
Comment (nicemicro) [reply godot_code/start_window.gd:34]: I guess we can just close the game window to simulate that kind of exit process.

File: controlpad/controller/controlpad.js (hunk 2 above)
Comment (nickydonna) [reply controlpad/controller/controlpad.js]: Yes I need some way to ge the information out. Right, at 
least on this code I don't have access to message, as far as I can tell

File: controlpad/controller/controlpad.js (hunk 3 above)
Comment (nickydonna) [reply controlpad/controller/controlpad.js]: I'll move this

**Review by safwaan-menk — APPROVED** [review]
I am approving to keep things going but it's critical to note that controlpad.js will definitely need to be updated as it needs to comply to the GameNite's controlpad.js standards.

So for now it's okay, but changes to controlpad.js will need to be made in a future update.

I will be going through the controlpad.js changes as there are good changes here and see what makes sense to incorporate and push it to the controlpad_test_server repo.  That controlpad.js must be used.

**Pull request merged**